# Dataset exploration

## import packages

In [33]:
import gzip
import json
from collections import defaultdict
import math
import scipy.optimize
from sklearn import svm
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import string
import random
import string
from sklearn import linear_model
from sklearn.metrics import accuracy_score
from scipy.optimize import fmin_l_bfgs_b

## Import Dataset

In [34]:
data_path = 'renttherunway_final_data.json.gz'

In [35]:
def readGz(path):
     g = gzip.open(path, 'r')
     for l in g:
        l = str(json.loads(l))
        yield eval(l)

In [36]:
data = []
for l in readGz(data_path):
    data.append(l)

In [52]:
data[0]

{'fit': 'fit',
 'user_id': '420272',
 'bust size': '34d',
 'item_id': '2260466',
 'weight': '137lbs',
 'rating': '10',
 'rented for': 'vacation',
 'review_text': "An adorable romper! Belt and zipper were a little hard to navigate in a full day of wear/bathroom use, but that's to be expected. Wish it had pockets, but other than that-- absolutely perfect! I got a million compliments.",
 'body type': 'hourglass',
 'review_summary': 'So many compliments!',
 'category': 'romper',
 'height': '5\' 8"',
 'size': 14,
 'age': '28',
 'review_date': 'April 20, 2016'}

In [74]:
df = pd.DataFrame(data)

In [75]:
display(df)

fit user_id bust size  item_id  weight rating     rented for  \
0       fit  420272       34d  2260466  137lbs     10       vacation   
1       fit  273551       34b   153475  132lbs     10          other   
2       fit  360448       NaN  1063761     NaN     10          party   
3       fit  909926       34c   126335  135lbs      8  formal affair   
4       fit  151944       34b   616682  145lbs     10        wedding   
...     ...     ...       ...      ...     ...    ...            ...   
192539  fit   66386      34dd  2252812  140lbs     10           work   
192540  fit  118398       32c   682043  100lbs     10           work   
192541  fit   47002       36a   683251  135lbs      6       everyday   
192542  fit  961120       36c   126335  165lbs     10        wedding   
192543  fit  123612       36b   127865  155lbs     10        wedding   

                                              review_text          body type  \
0       An adorable romper! Belt and zipper were a lit...          hourglass   
1       I rented this dress for a photo shoot. The the...  straight & narrow   
2       This hugged in all the right places! It was a ...                NaN   
3       I rented this for my company's black tie award...               pear   
4       I have always been petite in my upper body and...           athletic   
...                                                   ...                ...   
192539                                  Fit like a glove!          hourglass   
192540  The pattern contrast on this dress is really s...             petite   
192541  Like the other DVF wraps, the fit on this is f...  straight & narrow   
192542  This dress was PERFECTION.  it looked incredib...               pear   
192543  This dress was wonderful! I had originally pla...           athletic   

                                           review_summary  category height  \
0                                    So many compliments!    romper  5' 8"   
1                                 I felt so glamourous!!!      gown  5' 6"   
2       It was a great time to celebrate the (almost) ...    sheath  5' 4"   
3        Dress arrived on time and in perfect condition.      dress  5' 5"   
4                         Was in love with this dress !!!      gown  5' 9"   
...                                                   ...       ...    ...   
192539       LOVE IT!!! First Item Im thinking of buying!  jumpsuit  5' 9"   
192540                                           LOVE it!     dress  5' 1"   
192541                    Loud patterning, flattering fit     dress  5' 8"   
192542  loved this dress it was comfortable and photog...     dress  5' 6"   
192543  I wore this to a beautiful black tie optional ...      gown  5' 6"   

        size  age         review_date  
0         14   28      April 20, 2016  
1         12   36       June 18, 2013  
2          4  116   December 14, 2015  
3          8   34   February 12, 2014  
4         12   27  September 26, 2016  
...      ...  ...                 ...  
192539     8   42        May 18, 2016  
192540     4   29  September 30, 2016  
192541     8   31       March 4, 2016  
192542    16   31   November 25, 2015  
192543    16   30     August 29, 2017  

[192544 rows x 15 columns]

In [76]:
print("\nDataset info: \n")
print(df.info())


Dataset info: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 192544 entries, 0 to 192543
Data columns (total 15 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   fit             192544 non-null  object
 1   user_id         192544 non-null  object
 2   bust size       174133 non-null  object
 3   item_id         192544 non-null  object
 4   weight          162562 non-null  object
 5   rating          192462 non-null  object
 6   rented for      192534 non-null  object
 7   review_text     192544 non-null  object
 8   body type       177907 non-null  object
 9   review_summary  192544 non-null  object
 10  category        192544 non-null  object
 11  height          191867 non-null  object
 12  size            192544 non-null  int64 
 13  age             191584 non-null  object
 14  review_date     192544 non-null  object
dtypes: int64(1), object(14)
memory usage: 22.0+ MB
None


In [77]:
print("\nDataset null data sum: \n")
print(df.isnull().sum())


Dataset null data sum: 

fit                   0
user_id               0
bust size         18411
item_id               0
weight            29982
rating               82
rented for           10
review_text           0
body type         14637
review_summary        0
category              0
height              677
size                  0
age                 960
review_date           0
dtype: int64


## Dataset Description

The dataset we have chosen to work with contains measurements of clothing fit from RentTheRunaway, which is an e-commerce platform that allow users to rent and purchase clothings, accessories and more for work, special events and everyday wear. 

The total number of observations that have been gathered for this specific dataset is 192544, which we decided was an appropriate number of observations and enough interactions between useres and items to perform a correct data analysis and model building. 

The attributes of the dataset include:
| Column Name      | Data Type | # of Null | Description                                                                                               |
|------------------|-----------|-----------|-----------------------------------------------------------------------------------------------------------|
| fit              | object    | 0         | Indicates whether the fit was as expected, tight, or loose.                                               |
| user_id          | object    | 0         | Unique identifier for the user who submitted the review.                                                  |
| bust size        | object    | 18411     | The bust size of the user, typically given in formats like "34d".                                         |
| item_id          | object    | 0         | Unique identifier for the item being reviewed.                                                            |
| weight           | object    | 29982     | The weight of the user, typically in formats like "137lbs".                                               |
| rating           | object    | 82        | The rating given by the user for the item, typically on a scale of 1 to 10, incremented by 2                               |
| rented for       | object    | 10        | Purpose for which the item was rented, e.g., vacation, formal event, etc.                                 |
| review_text      | object    | 0         | The detailed text of the user's review about the item.                                                    |
| body type        | object    | 14637     | The body type of the user, e.g., hourglass, athletic, pear-shaped, etc.                                   |
| review_summary   | object    | 0         | A brief summary of the user's review.                                                                     |
| category         | object    | 0         | The category of the item being reviewed, e.g., romper, dress, etc.                                        |
| height           | object    | 677       | The height of the user, typically in formats like "5' 8\"".                                               |
| size             | int64     | 0         | The size of the item rented by the user, typically as a numeric value.                                    |
| age              | object    | 960       | The age of the user, typically in years.                                                                  |
| review_date      | object    | 0         | The date the review was submitted, typically in formats like "April 20, 2016".   



## Data cleanup and preprocessing

- 

In [78]:
def convert_weight(weight):
    if pd.isnull(weight):
        return None
    if isinstance(weight, float):
        return weight
    return float(weight.replace('lbs', '').strip())

In [79]:
def convert_height(height):
    if pd.isnull(height):
        return None
    if isinstance(height, float) or isinstance(height, int):
        return height
    feet, inches = height.split("'")
    feet = int(feet.strip())
    inches = int(inches.replace('"', '').strip())
    return feet * 12 + inches

In [80]:
def convert_to_float(x):
    if pd.isnull(x):
        return None
    try:
        return float(x)
    except (TypeError, ValueError): 
        return x

In [85]:
print(df['fit'].unique())

['fit' 'small' 'large']


In [84]:
print(df['bust size'].unique())

['34d' '34b' nan '34c' '32b' '34d+' '36d' '36a' '32d+' '36c' '32c' '32dd'
 '30dd' '32d' '36b' '34a' '34dd' '40ddd/e' '32a' '38d+' '38c' '34ddd/e'
 '38a' '38b' '38d' '36ddd/e' '38dd' '38f' '42dd' '36dd' '40g' '40dd' '28f'
 '30ddd/e' '32ddd/e' '28b' '36d+' '40d' '28dd' '34f' '42c' '30d' '30b'
 '42d' '34aa' '32aa' '36f' '28c' '40b' '36aa' '34g' '32f' '42ddd/e' '40c'
 '28g' '28aa' '36g' '30a' '44ddd/e' '30c' '30aa' '44dd' '32i' '32g' '30g'
 '38h' '30h' '34j' '38ddd/e' '32h' '28a' '44d' '40h' '38g' '34h' '36j'
 '40f' '36h' '28ddd/e' '44f' '34i' '30f' '36i' '30i' '28h' '42b' '42j'
 '32j' '46c' '48d' '38aa' '38i' '48dd' '28d' '28i' '46ddd/e' '42g' '44c'
 '42f' '40j' '38j' '46f' '44b' '44g' '42h' '40i' '44h']


In [81]:
df['weight'] = df['weight'].apply(convert_weight)
df['height'] = df['height'].apply(convert_height)
df['rating'] = df['rating'].apply(convert_to_float)
df['age'] = df['age'].apply(convert_to_float)

In [82]:
print(df.dtypes)

fit                object
user_id            object
bust size          object
item_id            object
weight            float64
rating            float64
rented for         object
review_text        object
body type          object
review_summary     object
category           object
height            float64
size                int64
age               float64
review_date        object
dtype: object


In [83]:
display(df.head())

fit user_id bust size  item_id  weight  rating     rented for  \
0  fit  420272       34d  2260466   137.0    10.0       vacation   
1  fit  273551       34b   153475   132.0    10.0          other   
2  fit  360448       NaN  1063761     NaN    10.0          party   
3  fit  909926       34c   126335   135.0     8.0  formal affair   
4  fit  151944       34b   616682   145.0    10.0        wedding   

                                         review_text          body type  \
0  An adorable romper! Belt and zipper were a lit...          hourglass   
1  I rented this dress for a photo shoot. The the...  straight & narrow   
2  This hugged in all the right places! It was a ...                NaN   
3  I rented this for my company's black tie award...               pear   
4  I have always been petite in my upper body and...           athletic   

                                      review_summary category  height  size  \
0                               So many compliments!   romper    68.0    14   
1                            I felt so glamourous!!!     gown    66.0    12   
2  It was a great time to celebrate the (almost) ...   sheath    64.0     4   
3   Dress arrived on time and in perfect condition.     dress    65.0     8   
4                    Was in love with this dress !!!     gown    69.0    12   

     age         review_date  
0   28.0      April 20, 2016  
1   36.0       June 18, 2013  
2  116.0   December 14, 2015  
3   34.0   February 12, 2014  
4   27.0  September 26, 2016

In [86]:
df.describe()

weight         rating         height           size  \
count  162562.000000  192462.000000  191867.000000  192544.000000   
mean      137.391709       9.092371      65.310621      12.245175   
std        21.899967       1.430044       2.663480       8.494877   
min        50.000000       2.000000      54.000000       0.000000   
25%       123.000000       8.000000      63.000000       8.000000   
50%       135.000000      10.000000      65.000000      12.000000   
75%       148.000000      10.000000      67.000000      16.000000   
max       300.000000      10.000000      78.000000      58.000000   

                 age  
count  191584.000000  
mean       33.871017  
std         8.058083  
min         0.000000  
25%        29.000000  
50%        32.000000  
75%        37.000000  
max       117.000000